core/embed/trezorhal/optiga/optiga.c

```diff
@@ -465,108 +530,150 @@ int optiga_pin_set(OPTIGA_UI_PROGRESS ui_progress,
-  res = optiga_pin_stretch_secret(ui_progress, out_secret);
+  // Authorise using OID_STRETCHED_PIN so that we can write to OID_PIN_HMAC and
+  // OID_PIN_HMAC_CTR.
+  res = optiga_set_auto_state(OPTIGA_OID_SESSION_CTX, OID_STRETCHED_PIN, digest,
+                              sizeof(digest));
+  memzero(digest, sizeof(digest));
   if (res != OPTIGA_SUCCESS) {
     return res;
   }
 
-  // Combine the stretched master secret with the PIN-derived secret to obtain
-  // the output secret. This ensures that in the unlikely case of an attacker
-  // recording communication between the MCU and Optiga, they cannot decrypt the
-  // storage without having to conduct a brute-force search for the PIN.
-  // NOTE: Considering how optiga_pin_stretch_secret() works internally and the
-  // fact that the PIN was already combined with the value of OID_PIN_SECRET,
-  // this step is not necessary. However, it is preferable to explicitly execute
-  // it, than to rely on the internals of optiga_pin_stretch_secret().
-  hmac_sha256(pin_secret, OPTIGA_PIN_SECRET_SIZE, out_secret,
-              OPTIGA_PIN_SECRET_SIZE, out_secret);
+  // Initialize the key for HMAC-SHA256 PIN stretching.
+  res = optiga_set_data_object(OID_PIN_HMAC, false, pin_hmac, sizeof(pin_hmac));
+  memzero(pin_hmac, sizeof(pin_hmac));
+  if (res != OPTIGA_SUCCESS) {
+    return res;
+  }
+
+  // Initialize the counter which limits the guesses at OID_STRETCHED_PIN again,
+  // since we just depleted one attempt.
+  res = optiga_set_data_object(OID_STRETCHED_PIN_CTR, false, COUNTER_RESET,
+                               sizeof(COUNTER_RESET));
+  optiga_clear_auto_state(OID_PIN_SECRET);
+  if (res != OPTIGA_SUCCESS) {
+    return res;
+  }
+
+  // Initialize the PIN counter which limits the use of OID_PIN_HMAC.
+  res = optiga_set_data_object(OID_PIN_HMAC_CTR, false, COUNTER_RESET,
+                               sizeof(COUNTER_RESET));
+  optiga_clear_auto_state(OID_STRETCHED_PIN);
+  if (res != OPTIGA_SUCCESS) {
+    return res;
+  }
+
+  ui_progress(250);
 
-  // Recombining the returned secret with the PIN-derived secret means that if
-  // the user chooses a high-entropy PIN, then even if the Optiga and its
-  // communication link is completely compromised, it will not reduce the
-  // security of their device any more than if the Optiga was not integrated
-  // into the device in the first place.
+  // Stretch the PIN more with the counter-protected PIN secret. This method
+  // ensures that if the user chooses a high-entropy PIN, then even if the
+  // Optiga and its communication link is completely compromised, it will not
+  // reduce the security of their device any more than if the Optiga was not
+  // integrated into the device in the first place.
+  hmac_sha256(stretched_pin, OPTIGA_PIN_SECRET_SIZE, pin_secret,
+              sizeof(pin_secret), stretched_pin);
+  memzero(pin_secret, sizeof(pin_secret));
+  if (res != OPTIGA_SUCCESS) {
+    return res;
+  }
```

Comment:
> Upon failure there is no need to clear them.

Yes, there is. You can have only four active authorizations. Clearing the authorizations helps prevent hard-to-find bugs in case we decide to add authorizations somewhere else in the future.

Reply:
If this function fails, then we have a major problem, because it brings the device into an inconsistent state from which it will probably not be able to recover in most cases. Indeed, that is why the PIN setting process is wrapped in `ensure()`.

Reply:
https://github.com/trezor/trezor-firmware/pull/3949/commits/3a6dbab0d9b8f43482a2ce372794aecaab394fb1